Get rid of goog.functions

bjornharrtell opened the pull request:
This PR suggests to remove `goog.functions` usage in the quest for closure library independence.

The reusable functions returning `true` or `false` have been ported to `ol.functions`. Other use of `goog.functions` had only a single call site and has been replaced by local code.

nicholas-l commented:
Haha awkward... #5084 should I close my one?

bjornharrtell commented:
Hah, yep. :) But there is actually value in this duplicated effort. I spot at least three things:

1. Good idea to have test cases, though they should be finished.
2. You are missing type annotations for some of the functions in `ol.functions`.
3. Ported code differs (at least in annotation). Should make sure code is ported from the correct closure library version (closure-util indicates version dated 20160208).

nicholas-l commented:
Yes, I am a bit new to type annotations.
One thing I did note was arguments usage, I am a bit unsure as to if this gets de-optimized. However  it is noted in [V8 Optimization Killers](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#what-is-safe-arguments-usage) that it can be if the function is bound and apply is used. Thats why I added the copying of arguments. Thoughts?

tschaub commented on src/ol/functions.js:
```diff
@@ -0,0 +1,70 @@
+ * @param {...Function} var_args A list of functions.
+ * @return {!Function} A function that calls all inputs in sequence.
+ */
+ol.functions.sequence = function(var_args) {
```
The one place this is used, we want to provide an array of functions.  I'd be tempted to use a `for` loop at the call site rather than define this and call it with `apply`.

tschaub commented on src/ol/functions.js:
```diff
@@ -0,0 +1,70 @@
+ * @return {function(...?):boolean} A function that ANDs its component
+ *      functions.
+ */
+ol.functions.and = function(var_args) {
```
Also only one call site for this.  Again tempting just to return an anonymous function where we use it.

sweco-sebhar commented:
@nicholas-l, I think we got good input (thanks @tschaub) to make a final effort on removing goog.functions usage. Do you want to do the honors or is it up to me? :)

nicholas-l commented:
Probably going to run out of time, grab anything from my pull request if you like

bjornharrtell commented:
I've made a new effort on this now and updated the PR description to the current status. I've replaced single code site usage with local code as recommended by @tschaub.

nicholas-l commented:
Looks great and now it doesn't look like it needs any tests added

bjornharrtell commented:
Rebased and ready for review. Ping @tschaub.

marcjansen commented on src/ol/style/atlasmanager.js:
```diff
@@ -186,7 +184,9 @@ ol.style.AtlasManager.prototype.add = function(id, width, height,
-      opt_renderHitCallback : goog.functions.NULL;
+      opt_renderHitCallback : function() {
+        return null;
+      };
```
Can we use `ol.nullFunction` here?

marcjansen commented:
I really like this. I added one minor comment / question.

Plus one for merging.

A follow-up PR might move [`ol.nullFunction`](https://github.com/openlayers/ol3/blob/master/src/ol/ol.js#L274) into the `ol.function` namespace.

Thanks for the work on this,  @bjornharrtell and @nicholas-l

bjornharrtell commented:
Thanks @marcjansen, wasn't aware of `ol.nullFunction`, updated branch to use it as suggested.

marcjansen commented:
Thanks. Can you rebase again, so I can merge this, @bjornharrtell?

bjornharrtell commented:
Rebased. Is it just me or was GitHub more lenient before, not requiring rebasing of out of date branches if they did not have merge conflicts?

marcjansen commented:
> Is it just me or was GitHub more lenient before, not requiring rebasing of out of date branches if they did not have merge conflicts?

It's not you alone… I think the same. It isn't that bad for everyone, though.

nicholas-l commented:
Could be to do with [this](https://help.github.com/articles/about-required-status-checks/)

bjornharrtell commented:
Ah nice find @nicholas-l. Should perhaps be discussed by the core devs. Problem is I don't know which forum, if any, to use to reach the core devs except here. @marcjansen any thoughts?

marcjansen commented:
Mhhm… maybe the dev mailing list or a dedicated issue here with a good rationale and some suggestions.

In this ticket here, the context is obviously wrong :wink:

probins commented:
Github have just changed the way this works when you have status checks https://github.com/blog/2137-protected-branches-improvements
